src/renderer/components/PerCurrencySelectAccount/state.js

```diff
@@ -0,0 +1,120 @@
+// @flow
+
+import { useState, useCallback, useMemo } from "react";
+import type { Account, SubAccount } from "@ledgerhq/live-common/lib/types/account";
+import type { CryptoCurrency, TokenCurrency } from "@ledgerhq/live-common/lib/types/currencies";
+import { BigNumber } from "bignumber.js";
+
+type CryptoOrTokenCurrency = TokenCurrency | CryptoCurrency;
+
+const generateTokenAccount = (account: Account, currency: TokenCurrency): SubAccount => ({
+  type: "TokenAccount",
+  id: account.id + "+" + currency.contractAddress,
+  parentId: account.id,
+  token: currency,
+  balance: BigNumber(0),
+  operationsCount: 0,
+  creationDate: new Date(),
+  operations: [],
+  pendingOperations: [],
+  starred: false,
+});
+
+export type AccountTuple = {
+  account: ?Account,
+  subAccount: ?SubAccount,
+};
+
+function getAccountTuplesForCurrency(
+  currency: CryptoOrTokenCurrency,
+  allAccounts: Account[],
+): AccountTuple[] {
+  if (currency.type === "TokenCurrency") {
+    return allAccounts
+      .filter(account => account.currency.id === currency.parentCurrency.id)
+      .map(account => ({
+        account,
+        subAccount:
+          (account.subAccounts &&
+            account.subAccounts.find(
+              (subAcc: SubAccount) =>
+                subAcc.type === "TokenAccount" && subAcc.token.id === currency.id,
+            )) ||
+          generateTokenAccount(account, currency),
+      }));
+  }
+  return allAccounts
+    .filter(account => account.currency.id === currency.id)
+    .map(account => ({
+      account,
+      subAccount: null,
+    }));
+}
+
+export function useCurrencyAccountSelect(
+  allCurrencies: CryptoOrTokenCurrency[],
+  allAccounts: Account[],
+) {
+  const [state, setState] = useState(() => {
+    const defaultCurrency = allCurrencies[0];
+    const availableAccounts = getAccountTuplesForCurrency(defaultCurrency, allAccounts);
+    const { account, subAccount } = availableAccounts.length
+      ? availableAccounts[0]
+      : { account: null, subAccount: null };
+
+    return {
+      currency: defaultCurrency,
+      account,
+      subAccount,
+    };
+  });
+
+  const { currency, account, subAccount } = state;
```

Comment:
when looking at the state currently involved, i think there can still be a case where allAccounts is more updated than the internal state. I mean, there is clearly two single source of trust regarding the reference to account

a solution to this problem is to do this instead:

```js
const account = allAccounts.find(a => a.id===accountId)
const subAccount = (account && account.subAccounts || []).find(a => a.id===subAccountId)
```

where `accountId` and `subAccountId` would be what is actually stored as a state in this hook.

in case of Buy feature, you probably didn't see any issue because we don't care about account updates & it doesn't impact the feature. In case of swap, it's way more dangerous (e.g. balance can updates)

Reply:
Yep but what if the subaccount doesn't actually exist ?